faster TensorOptions merging

bhosmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45263 use bitfield to shrink TensorImpl
* **#45046 faster TensorOptions merging**
* #45043 track Half/ComplexHalf default dtype
* #45026 pass TypeMeta by value

Turns out `TensorOptions::merge_in` is pretty hot - 2-3% of callgrind IR in an empty() microbenchmark. This PR does a couple simple things:
* rewrites it to be fastest when RHS has fewest contributions to the merge. Previously it was fast when left (i.e., `this`) had fewest, which in practice is less common. No change in output, obviously.
  * **Note**: this change is actually the least significant contributor to the perf improvement, so if there's any subtlety around its original implementation we could just restore it without losing the bulk of the gains
* adds the method `merge_memory_format()`. This can be removed once TensorOptions has been rationalized out of kernel signatures, but in the meantime it saves a surprising amount of work
* light refactoring of one of the hot spots to avoid a merge

Saves like ~2% IR on empty() microbenchmark

Differential Revision: [D23806787](https://our.internmc.facebook.com/intern/diff/D23806787)

Commit:
faster TensorOptions merging

[ghstack-poisoned]

Commit:
Update on "faster TensorOptions merging"


Turns out `TensorOptions::merge_in` is pretty hot - 2-3% of callgrind IR in an empty() microbenchmark. This PR does a couple simple things:
* rewrites it to be fastest when RHS has fewest contributions to the merge. Previously it was fast when left (i.e., `this`) had fewest, which in practice is less common. No change in output, obviously
* adds the method `merge_memory_format()`. This can be removed once TensorOptions has been rationalized out of kernel signatures, but in the meantime it saves a surprising amount of work
* light refactoring of one of the hot spots to avoid a merge

Saves like ~2% IR on empty() microbenchmark

Differential Revision: [D23806787](https://our.internmc.facebook.com/intern/diff/D23806787)

[ghstack-poisoned]

Commit:
Update on "faster TensorOptions merging"


Turns out `TensorOptions::merge_in` is pretty hot - 2-3% of callgrind IR in an empty() microbenchmark. This PR does a couple simple things:
* rewrites it to be fastest when RHS has fewest contributions to the merge. Previously it was fast when left (i.e., `this`) had fewest, which in practice is less common. No change in output, obviously.
  * **Note**: this change is actually the least significant contributor to the perf improvement, so if there's any subtlety around its original implementation we could just restore it without losing the bulk of the gains
* adds the method `merge_memory_format()`. This can be removed once TensorOptions has been rationalized out of kernel signatures, but in the meantime it saves a surprising amount of work
* light refactoring of one of the hot spots to avoid a merge

Saves like ~2% IR on empty() microbenchmark

Differential Revision: [D23806787](https://our.internmc.facebook.com/intern/diff/D23806787)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/TensorConversions.cpp:
```diff
@@ -56,7 +56,7 @@ Tensor to(
     "Cannot set memory_format both in TensorOptions and explicit argument; please delete "
     "the redundant setter.");
-  auto options = options_.merge_in(TensorOptions().memory_format(optional_memory_format));
+  auto options = options_.merge_memory_format(optional_memory_format);
```
Drat! It's too bad inlining can't figure it out lol

ezyang commented on c10/core/TensorOptions.h:
```diff
@@ -528,7 +537,7 @@ struct C10_API TensorOptions {
   // the has_***_ boolean fields.
 
-  caffe2::TypeMeta dtype_ = caffe2::TypeMeta::Make<float>(); // 64-bit
+  caffe2::TypeMeta dtype_ = caffe2::TypeMeta::Make<float>(); // 16-bit
```
Now that it's 16-bit it will be profitable to move this field down one, so it can get coalesced with layout and memory format.

bhosmer replied:
Yessss

Commit:
Update on "faster TensorOptions merging"


Turns out `TensorOptions::merge_in` is pretty hot - 2-3% of callgrind IR in an empty() microbenchmark. This PR does a couple simple things:
* rewrites it to be fastest when RHS has fewest contributions to the merge. Previously it was fast when left (i.e., `this`) had fewest, which in practice is less common. No change in output, obviously.
  * **Note**: this change is actually the least significant contributor to the perf improvement, so if there's any subtlety around its original implementation we could just restore it without losing the bulk of the gains
* adds the method `merge_memory_format()`. This can be removed once TensorOptions has been rationalized out of kernel signatures, but in the meantime it saves a surprising amount of work
* light refactoring of one of the hot spots to avoid a merge

Saves like ~2% IR on empty() microbenchmark

Differential Revision: [D23806787](https://our.internmc.facebook.com/intern/diff/D23806787)

[ghstack-poisoned]

Commit:
Update on "faster TensorOptions merging"


Turns out `TensorOptions::merge_in` is pretty hot - 2-3% of callgrind IR in an empty() microbenchmark. This PR does a couple simple things:
* rewrites it to be fastest when RHS has fewest contributions to the merge. Previously it was fast when left (i.e., `this`) had fewest, which in practice is less common. No change in output, obviously.
  * **Note**: this change is actually the least significant contributor to the perf improvement, so if there's any subtlety around its original implementation we could just restore it without losing the bulk of the gains
* adds the method `merge_memory_format()`. This can be removed once TensorOptions has been rationalized out of kernel signatures, but in the meantime it saves a surprising amount of work
* light refactoring of one of the hot spots to avoid a merge

Saves like ~2% IR on empty() microbenchmark

Differential Revision: [D23806787](https://our.internmc.facebook.com/intern/diff/D23806787)

[ghstack-poisoned]

Commit:
Update on "faster TensorOptions merging"


Turns out `TensorOptions::merge_in` is pretty hot - 2-3% of callgrind IR in an empty() microbenchmark. This PR does a couple simple things:
* rewrites it to be fastest when RHS has fewest contributions to the merge. Previously it was fast when left (i.e., `this`) had fewest, which in practice is less common. No change in output, obviously.
  * **Note**: this change is actually the least significant contributor to the perf improvement, so if there's any subtlety around its original implementation we could just restore it without losing the bulk of the gains
* adds the method `merge_memory_format()`. This can be removed once TensorOptions has been rationalized out of kernel signatures, but in the meantime it saves a surprising amount of work
* light refactoring of one of the hot spots to avoid a merge

Saves like ~2% IR on empty() microbenchmark

Differential Revision: [D23806787](https://our.internmc.facebook.com/intern/diff/D23806787)

[ghstack-poisoned]

Commit:
Update on "faster TensorOptions merging"


Turns out `TensorOptions::merge_in` is pretty hot - 2-3% of callgrind IR in an empty() microbenchmark. This PR does a couple simple things:
* rewrites it to be fastest when RHS has fewest contributions to the merge. Previously it was fast when left (i.e., `this`) had fewest, which in practice is less common. No change in output, obviously.
  * **Note**: this change is actually the least significant contributor to the perf improvement, so if there's any subtlety around its original implementation we could just restore it without losing the bulk of the gains
* adds the method `merge_memory_format()`. This can be removed once TensorOptions has been rationalized out of kernel signatures, but in the meantime it saves a surprising amount of work
* light refactoring of one of the hot spots to avoid a merge

Saves like ~2% IR on empty() microbenchmark

Differential Revision: [D23806787](https://our.internmc.facebook.com/intern/diff/D23806787)

[ghstack-poisoned]

Commit:
Update on "faster TensorOptions merging"


Turns out `TensorOptions::merge_in` is pretty hot - 2-3% of callgrind IR in an empty() microbenchmark. This PR does a couple simple things:
* rewrites it to be fastest when RHS has fewest contributions to the merge. Previously it was fast when left (i.e., `this`) had fewest, which in practice is less common. No change in output, obviously.
  * **Note**: this change is actually the least significant contributor to the perf improvement, so if there's any subtlety around its original implementation we could just restore it without losing the bulk of the gains
* adds the method `merge_memory_format()`. This can be removed once TensorOptions has been rationalized out of kernel signatures, but in the meantime it saves a surprising amount of work
* light refactoring of one of the hot spots to avoid a merge

Saves like ~2% IR on empty() microbenchmark

Differential Revision: [D23806787](https://our.internmc.facebook.com/intern/diff/D23806787)

[ghstack-poisoned]

Commit:
Update on "faster TensorOptions merging"


Turns out `TensorOptions::merge_in` is pretty hot - 2-3% of callgrind IR in an empty() microbenchmark. This PR does a couple simple things:
* rewrites it to be fastest when RHS has fewest contributions to the merge. Previously it was fast when left (i.e., `this`) had fewest, which in practice is less common. No change in output, obviously.
  * **Note**: this change is actually the least significant contributor to the perf improvement, so if there's any subtlety around its original implementation we could just restore it without losing the bulk of the gains
* adds the method `merge_memory_format()`. This can be removed once TensorOptions has been rationalized out of kernel signatures, but in the meantime it saves a surprising amount of work
* light refactoring of one of the hot spots to avoid a merge

Saves like ~2% IR on empty() microbenchmark

Differential Revision: [D23806787](https://our.internmc.facebook.com/intern/diff/D23806787)

[ghstack-poisoned]

Commit:
Update on "faster TensorOptions merging"


Turns out `TensorOptions::merge_in` is pretty hot - 2-3% of callgrind IR in an empty() microbenchmark. This PR does a couple simple things:
* rewrites it to be fastest when RHS has fewest contributions to the merge. Previously it was fast when left (i.e., `this`) had fewest, which in practice is less common. No change in output, obviously.
  * **Note**: this change is actually the least significant contributor to the perf improvement, so if there's any subtlety around its original implementation we could just restore it without losing the bulk of the gains
* adds the method `merge_memory_format()`. This can be removed once TensorOptions has been rationalized out of kernel signatures, but in the meantime it saves a surprising amount of work
* light refactoring of one of the hot spots to avoid a merge

Saves like ~2% IR on empty() microbenchmark

Differential Revision: [D23806787](https://our.internmc.facebook.com/intern/diff/D23806787)

[ghstack-poisoned]

Commit:
Update on "faster TensorOptions merging"


Turns out `TensorOptions::merge_in` is pretty hot - 2-3% of callgrind IR in an empty() microbenchmark. This PR does a couple simple things:
* rewrites it to be fastest when RHS has fewest contributions to the merge. Previously it was fast when left (i.e., `this`) had fewest, which in practice is less common. No change in output, obviously.
  * **Note**: this change is actually the least significant contributor to the perf improvement, so if there's any subtlety around its original implementation we could just restore it without losing the bulk of the gains
* adds the method `merge_memory_format()`. This can be removed once TensorOptions has been rationalized out of kernel signatures, but in the meantime it saves a surprising amount of work
* light refactoring of one of the hot spots to avoid a merge

Saves like ~2% IR on empty() microbenchmark

Differential Revision: [D23806787](https://our.internmc.facebook.com/intern/diff/D23806787)

[ghstack-poisoned]

Commit:
Update on "faster TensorOptions merging"


Turns out `TensorOptions::merge_in` is pretty hot - 2-3% of callgrind IR in an empty() microbenchmark. This PR does a couple simple things:
* rewrites it to be fastest when RHS has fewest contributions to the merge. Previously it was fast when left (i.e., `this`) had fewest, which in practice is less common. No change in output, obviously.
  * **Note**: this change is actually the least significant contributor to the perf improvement, so if there's any subtlety around its original implementation we could just restore it without losing the bulk of the gains
* adds the method `merge_memory_format()`. This can be removed once TensorOptions has been rationalized out of kernel signatures, but in the meantime it saves a surprising amount of work
* light refactoring of one of the hot spots to avoid a merge

Saves like ~2% IR on empty() microbenchmark

Differential Revision: [D23806787](https://our.internmc.facebook.com/intern/diff/D23806787)

[ghstack-poisoned]

Commit:
Update on "faster TensorOptions merging"


Turns out `TensorOptions::merge_in` is pretty hot - 2-3% of callgrind IR in an empty() microbenchmark. This PR does a couple simple things:
* rewrites it to be fastest when RHS has fewest contributions to the merge. Previously it was fast when left (i.e., `this`) had fewest, which in practice is less common. No change in output, obviously.
  * **Note**: this change is actually the least significant contributor to the perf improvement, so if there's any subtlety around its original implementation we could just restore it without losing the bulk of the gains
* adds the method `merge_memory_format()`. This can be removed once TensorOptions has been rationalized out of kernel signatures, but in the meantime it saves a surprising amount of work
* light refactoring of one of the hot spots to avoid a merge

Saves like ~2% IR on empty() microbenchmark

Differential Revision: [D23806787](https://our.internmc.facebook.com/intern/diff/D23806787)

[ghstack-poisoned]

Commit:
Update on "faster TensorOptions merging"


Turns out `TensorOptions::merge_in` is pretty hot - 2-3% of callgrind IR in an empty() microbenchmark. This PR does a couple simple things:
* rewrites it to be fastest when RHS has fewest contributions to the merge. Previously it was fast when left (i.e., `this`) had fewest, which in practice is less common. No change in output, obviously.
  * **Note**: this change is actually the least significant contributor to the perf improvement, so if there's any subtlety around its original implementation we could just restore it without losing the bulk of the gains
* adds the method `merge_memory_format()`. This can be removed once TensorOptions has been rationalized out of kernel signatures, but in the meantime it saves a surprising amount of work
* light refactoring of one of the hot spots to avoid a merge

Saves like ~2% IR on empty() microbenchmark

Differential Revision: [D23806787](https://our.internmc.facebook.com/intern/diff/D23806787)

[ghstack-poisoned]

Commit:
Update on "faster TensorOptions merging"


Turns out `TensorOptions::merge_in` is pretty hot - 2-3% of callgrind IR in an empty() microbenchmark. This PR does a couple simple things:
* rewrites it to be fastest when RHS has fewest contributions to the merge. Previously it was fast when left (i.e., `this`) had fewest, which in practice is less common. No change in output, obviously.
  * **Note**: this change is actually the least significant contributor to the perf improvement, so if there's any subtlety around its original implementation we could just restore it without losing the bulk of the gains
* adds the method `merge_memory_format()`. This can be removed once TensorOptions has been rationalized out of kernel signatures, but in the meantime it saves a surprising amount of work
* light refactoring of one of the hot spots to avoid a merge

Saves like ~2% IR on empty() microbenchmark

Differential Revision: [D23806787](https://our.internmc.facebook.com/intern/diff/D23806787)

[ghstack-poisoned]

Commit:
Update on "faster TensorOptions merging"


Turns out `TensorOptions::merge_in` is pretty hot - 2-3% of callgrind IR in an empty() microbenchmark. This PR does a couple simple things:
* rewrites it to be fastest when RHS has fewest contributions to the merge. Previously it was fast when left (i.e., `this`) had fewest, which in practice is less common. No change in output, obviously.
  * **Note**: this change is actually the least significant contributor to the perf improvement, so if there's any subtlety around its original implementation we could just restore it without losing the bulk of the gains
* adds the method `merge_memory_format()`. This can be removed once TensorOptions has been rationalized out of kernel signatures, but in the meantime it saves a surprising amount of work
* light refactoring of one of the hot spots to avoid a merge

Saves like ~2% IR on empty() microbenchmark

Differential Revision: [D23806787](https://our.internmc.facebook.com/intern/diff/D23806787)

[ghstack-poisoned]

Commit:
Update on "faster TensorOptions merging"


Turns out `TensorOptions::merge_in` is pretty hot - 2-3% of callgrind IR in an empty() microbenchmark. This PR does a couple simple things:
* rewrites it to be fastest when RHS has fewest contributions to the merge. Previously it was fast when left (i.e., `this`) had fewest, which in practice is less common. No change in output, obviously.
  * **Note**: this change is actually the least significant contributor to the perf improvement, so if there's any subtlety around its original implementation we could just restore it without losing the bulk of the gains
* adds the method `merge_memory_format()`. This can be removed once TensorOptions has been rationalized out of kernel signatures, but in the meantime it saves a surprising amount of work
* light refactoring of one of the hot spots to avoid a merge

Saves like ~2% IR on empty() microbenchmark

Differential Revision: [D23806787](https://our.internmc.facebook.com/intern/diff/D23806787)

[ghstack-poisoned]

Commit:
Update on "faster TensorOptions merging"


Turns out `TensorOptions::merge_in` is pretty hot - 2-3% of callgrind IR in an empty() microbenchmark. This PR does a couple simple things:
* rewrites it to be fastest when RHS has fewest contributions to the merge. Previously it was fast when left (i.e., `this`) had fewest, which in practice is less common. No change in output, obviously.
  * **Note**: this change is actually the least significant contributor to the perf improvement, so if there's any subtlety around its original implementation we could just restore it without losing the bulk of the gains
* adds the method `merge_memory_format()`. This can be removed once TensorOptions has been rationalized out of kernel signatures, but in the meantime it saves a surprising amount of work
* light refactoring of one of the hot spots to avoid a merge

Saves like ~2% IR on empty() microbenchmark

Differential Revision: [D23806787](https://our.internmc.facebook.com/intern/diff/D23806787)

[ghstack-poisoned]

Commit:
Update on "faster TensorOptions merging"


Turns out `TensorOptions::merge_in` is pretty hot - 2-3% of callgrind IR in an empty() microbenchmark. This PR does a couple simple things:
* rewrites it to be fastest when RHS has fewest contributions to the merge. Previously it was fast when left (i.e., `this`) had fewest, which in practice is less common. No change in output, obviously.
  * **Note**: this change is actually the least significant contributor to the perf improvement, so if there's any subtlety around its original implementation we could just restore it without losing the bulk of the gains
* adds the method `merge_memory_format()`. This can be removed once TensorOptions has been rationalized out of kernel signatures, but in the meantime it saves a surprising amount of work
* light refactoring of one of the hot spots to avoid a merge

Saves like ~2% IR on empty() microbenchmark

Differential Revision: [D23806787](https://our.internmc.facebook.com/intern/diff/D23806787)

[ghstack-poisoned]

Commit:
Update on "faster TensorOptions merging"


Turns out `TensorOptions::merge_in` is pretty hot - 2-3% of callgrind IR in an empty() microbenchmark. This PR does a couple simple things:
* rewrites it to be fastest when RHS has fewest contributions to the merge. Previously it was fast when left (i.e., `this`) had fewest, which in practice is less common. No change in output, obviously.
  * **Note**: this change is actually the least significant contributor to the perf improvement, so if there's any subtlety around its original implementation we could just restore it without losing the bulk of the gains
* adds the method `merge_memory_format()`. This can be removed once TensorOptions has been rationalized out of kernel signatures, but in the meantime it saves a surprising amount of work
* light refactoring of one of the hot spots to avoid a merge

Saves like ~2% IR on empty() microbenchmark

Differential Revision: [D23806787](https://our.internmc.facebook.com/intern/diff/D23806787)

[ghstack-poisoned]

Commit:
Update on "faster TensorOptions merging"


Turns out `TensorOptions::merge_in` is pretty hot - 2-3% of callgrind IR in an empty() microbenchmark. This PR does a couple simple things:
* rewrites it to be fastest when RHS has fewest contributions to the merge. Previously it was fast when left (i.e., `this`) had fewest, which in practice is less common. No change in output, obviously.
  * **Note**: this change is actually the least significant contributor to the perf improvement, so if there's any subtlety around its original implementation we could just restore it without losing the bulk of the gains
* adds the method `merge_memory_format()`. This can be removed once TensorOptions has been rationalized out of kernel signatures, but in the meantime it saves a surprising amount of work
* light refactoring of one of the hot spots to avoid a merge

Saves like ~2% IR on empty() microbenchmark

Differential Revision: [D23806787](https://our.internmc.facebook.com/intern/diff/D23806787)

[ghstack-poisoned]

Commit:
Update on "faster TensorOptions merging"


Turns out `TensorOptions::merge_in` is pretty hot - 2-3% of callgrind IR in an empty() microbenchmark. This PR does a couple simple things:
* rewrites it to be fastest when RHS has fewest contributions to the merge. Previously it was fast when left (i.e., `this`) had fewest, which in practice is less common. No change in output, obviously.
  * **Note**: this change is actually the least significant contributor to the perf improvement, so if there's any subtlety around its original implementation we could just restore it without losing the bulk of the gains
* adds the method `merge_memory_format()`. This can be removed once TensorOptions has been rationalized out of kernel signatures, but in the meantime it saves a surprising amount of work
* light refactoring of one of the hot spots to avoid a merge

Saves like ~2% IR on empty() microbenchmark

Differential Revision: [D23806787](https://our.internmc.facebook.com/intern/diff/D23806787)

[ghstack-poisoned]

Commit:
Update on "faster TensorOptions merging"


Turns out `TensorOptions::merge_in` is pretty hot - 2-3% of callgrind IR in an empty() microbenchmark. This PR does a couple simple things:
* rewrites it to be fastest when RHS has fewest contributions to the merge. Previously it was fast when left (i.e., `this`) had fewest, which in practice is less common. No change in output, obviously.
  * **Note**: this change is actually the least significant contributor to the perf improvement, so if there's any subtlety around its original implementation we could just restore it without losing the bulk of the gains
* adds the method `merge_memory_format()`. This can be removed once TensorOptions has been rationalized out of kernel signatures, but in the meantime it saves a surprising amount of work
* light refactoring of one of the hot spots to avoid a merge

Saves like ~2% IR on empty() microbenchmark

Differential Revision: [D23806787](https://our.internmc.facebook.com/intern/diff/D23806787)

[ghstack-poisoned]

Commit:
Update on "faster TensorOptions merging"


Turns out `TensorOptions::merge_in` is pretty hot - 2-3% of callgrind IR in an empty() microbenchmark. This PR does a couple simple things:
* rewrites it to be fastest when RHS has fewest contributions to the merge. Previously it was fast when left (i.e., `this`) had fewest, which in practice is less common. No change in output, obviously.
  * **Note**: this change is actually the least significant contributor to the perf improvement, so if there's any subtlety around its original implementation we could just restore it without losing the bulk of the gains
* adds the method `merge_memory_format()`. This can be removed once TensorOptions has been rationalized out of kernel signatures, but in the meantime it saves a surprising amount of work
* light refactoring of one of the hot spots to avoid a merge

Saves like ~2% IR on empty() microbenchmark

Differential Revision: [D23806787](https://our.internmc.facebook.com/intern/diff/D23806787)

[ghstack-poisoned]

Commit:
Update on "faster TensorOptions merging"


Turns out `TensorOptions::merge_in` is pretty hot - 2-3% of callgrind IR in an empty() microbenchmark. This PR does a couple simple things:
* rewrites it to be fastest when RHS has fewest contributions to the merge. Previously it was fast when left (i.e., `this`) had fewest, which in practice is less common. No change in output, obviously.
  * **Note**: this change is actually the least significant contributor to the perf improvement, so if there's any subtlety around its original implementation we could just restore it without losing the bulk of the gains
* adds the method `merge_memory_format()`. This can be removed once TensorOptions has been rationalized out of kernel signatures, but in the meantime it saves a surprising amount of work
* light refactoring of one of the hot spots to avoid a merge

Saves like ~2% IR on empty() microbenchmark

Differential Revision: [D23806787](https://our.internmc.facebook.com/intern/diff/D23806787)

[ghstack-poisoned]

Commit:
Update on "faster TensorOptions merging"


Turns out `TensorOptions::merge_in` is pretty hot - 2-3% of callgrind IR in an empty() microbenchmark. This PR does a couple simple things:
* rewrites it to be fastest when RHS has fewest contributions to the merge. Previously it was fast when left (i.e., `this`) had fewest, which in practice is less common. No change in output, obviously.
  * **Note**: this change is actually the least significant contributor to the perf improvement, so if there's any subtlety around its original implementation we could just restore it without losing the bulk of the gains
* adds the method `merge_memory_format()`. This can be removed once TensorOptions has been rationalized out of kernel signatures, but in the meantime it saves a surprising amount of work
* light refactoring of one of the hot spots to avoid a merge

Saves like ~2% IR on empty() microbenchmark

Differential Revision: [D23806787](https://our.internmc.facebook.com/intern/diff/D23806787)

[ghstack-poisoned]

Commit:
Update on "faster TensorOptions merging"


Turns out `TensorOptions::merge_in` is pretty hot - 2-3% of callgrind IR in an empty() microbenchmark. This PR does a couple simple things:
* rewrites it to be fastest when RHS has fewest contributions to the merge. Previously it was fast when left (i.e., `this`) had fewest, which in practice is less common. No change in output, obviously.
  * **Note**: this change is actually the least significant contributor to the perf improvement, so if there's any subtlety around its original implementation we could just restore it without losing the bulk of the gains
* adds the method `merge_memory_format()`. This can be removed once TensorOptions has been rationalized out of kernel signatures, but in the meantime it saves a surprising amount of work
* light refactoring of one of the hot spots to avoid a merge

Saves like ~2% IR on empty() microbenchmark

Differential Revision: [D23806787](https://our.internmc.facebook.com/intern/diff/D23806787)

[ghstack-poisoned]
